### DISCUSSION TITLE
How do debug decisions made by composer update (-w/-W)

### DISCUSSION THREAD

**Discussion by mxr576** [composer/composer#10093] [Q&A / Support] [answered]
Is there a way to debug decisions made by composer update (-w/-W)? It keeps updating packages and even if I think I understand how -w/-W suppose to work, I do not understand why it updates some packages on a project.

![image](https://user-images.githubusercontent.com/1755573/132658168-d4932645-db67-4160-8cc1-b65cba09eab7.png)

Also, am I the only one who does not understand why Composer (2) updates packages? (Am I the only one who is concerned by unexpected package updates? :P ) 

Composer version: 2.1.6

**Comment by mxr576** [comment]
I was curious so I checked if the same happens with Composer 1. I did not understand why `apigee/apigee-client-php` gets updated with `-w` and `-W`. Interestingly, Composer 1 does not want to update that package on the same project (and yes, Composer 1 still has access to the latest releases of that package, see output.).

[composer1.txt](https://github.com/composer/composer/files/7135646/composer1.txt)
[composer2.txt](https://github.com/composer/composer/files/7135647/composer2.txt)

Unfortunately, the project is internal, packages are internal, so I cannot share that. Not sure how much time would it take to create a minimum code where that can be reproduced... so if you experience the same and you can share a project where it happens, please share it here.

**Reply by stof** [reply]
Without seeing the dependency tree of the pronovix packages that are used to trigger the update, there is nothing we can do to debug this

**Reply by mxr576** [reply]
Right, how can I debug this myself? This was the original conversation starter. Because I used `composer why apigee/apigee-client-php` to double-check and none of the Pronovix packages are depending on that library. Also checked conflicts and minimum requirements in `apigee/apigee-client-php:2.0.8` to confirm that none of the updated packages requires that component to be also updated. So technically speaking, I do not see a reason why Composer 2 updates it... and it is quite interesting that Composer 1 does not. Or what else I can check to see the "WHY"? 

Maybe with `composer show --tree`?

**Reply by stof** [reply]
try to see `composer why apigee/apigee-client-php -r` to display it in recursive mode. Maybe one of the dependencies of a pronovix package depend on this apigee package.
`composer show --tree` might also allow you to find the info, but it will be harder as you will have a lot more noise.

**Reply by mxr576** [reply]
`composer why apigee/apigee-client-php -r` only returns the root composer and other project-specific local packages. The root composer.json ofc depends on pronovix/* packages in `require`, project-specific local packages only depend on pronovix/* packages in `require-dev`. Both C1 and C2 returns the same result.
`composer show --tree` basically confirms the same as I see.

**Reply by Seldaek** [reply]
One option I see is that you have `apigee/apigee-client-php` defined as `path` repository, because v2 always updates those. If not, then maybe it's due to the conflicts on guzzlehttp packages that it gets selected, if something depends on guzzlehttp?

Anyway the bottom line is.. either you trust Composer is doing the right thing, or if you don't like it you can use `composer update provonix/* symfony/* php-http/* nikic/php-parser nette/utils instaclick/php-webdriver` etc as needed to make sure the update succeeds without -w/-W.

**Reply by mxr576** [reply]
> One option I see is that you have apigee/apigee-client-php defined as path repository, because v2 always updates those. If

Hm, `apigee/apigee-client-php` indeed a dependency of a local package that is being installed in root composer.json. (Included via `path` repository.) So Composer 2 always updates local packages even if they do not match to the `pronovix/*` package name pattern...   and because `-w/-W` is used, also their dependencies too? :roll_eyes:  Am I understand this right? What was the motivation behind adding this new behavior?

Is there an option already to disable updating packages in a path repository for a time being? If not, something like this could work: `composer update pronovix/* -W --no-local-package-update`


Update: meanwhile I confirmed, if no local package requires `apigee/apigee-client-php` just for example, the root composer.json, it does not get updated.

**Reply by mxr576** [reply]
@Seldaek hi, can you share some information related to this question? Otherwise, I consider this is resolved, I just wanted to learn about how path repository behaviors have changed in Composer 2 since I am extensively using it in our projects.

> What was the motivation behind adding this new behavior?

**Reply by stof** [reply]
@mxr576 as path repositories are symlinking the code, the content of the package is actually already updating even without running a `composer update` at all. Updating them on each update even when not whitelisted explicitly was a way to use the right package metadata (list of dependencies, autoloading rules) corresponding to that source code at least when you run a `composer update`.

@Seldaek maybe this forced inclusion of path packages should be done *after* building the whitelist with `-w/-W`, so that the path packages don't automatically whitelist their dependencies when this option are used (they would still whitelist their dependencies if the path packages are explicitly marked for update in the whitelist in the command)

**Reply by mxr576** [reply]
> maybe this forced inclusion of path packages should be done after building the whitelist with -w/-W, so that the path packages don't automatically whitelist their dependencies when this option is used (they would still whitelist their dependencies if the path packages are explicitly marked for update in the whitelist in the command)

:heavy_plus_sign:  on this, should we open a new issue about this?

**Reply by stof** [reply]
@mxr576 that makes sense

**Reply by Seldaek** [reply]
See https://github.com/composer/composer/pull/10157

**Reply by mxr576** [reply]
thanks @Seldaek, this was still in my todo list

**Comment by mxr576** [comment, marked as answer]
I consider this as resolved.

https://github.com/composer/composer/discussions/10093#discussioncomment-1374547 solves my original problem and it seems the answer to my original question is: there is no way to debug it.
